packages/core/contracts/insured-bridge/BridgePool.sol

```diff
@@ -388,13 +383,60 @@ contract BridgePool is Testable, BridgePoolInterface, ExpandedERC20, MultiCaller
         updateAccumulatedLpFees();
         allocateLpFees(_getAmountFromPct(relay.realizedLpFeePct, _depositData.amount));
 
-        emit RelaySettled(depositHash, relayHash, msg.sender);
+        emit RelaySettled(depositHash, _getRelayHash(_depositData, relay), msg.sender);
 
         delete instantRelays[keccak256(abi.encode(depositHash, relay.realizedLpFeePct))];
         delete relay.realizedLpFeePct;
         delete relay.priceRequestTime;
     }
 
+    /**
+     * @notice Callback for disputes, marks relay as disputed.
+     * @dev Reverts if relay is not pending.
+     * @dev _timestamp and _identifier are unused because _ancillaryData contains a relay nonce and uniquely
+     * identifies a relay request.
+     * @param _identifier price identifier for relay request.
+     * @param _timestamp timestamp for relay request.
+     * @param _ancillaryData ancillary data for relay request.
+     * @param _request disputed relay request params.
+     */
+    function priceDisputed(
+        bytes32 _identifier,
+        uint32 _timestamp,
+        bytes memory _ancillaryData,
+        SkinnyOptimisticOracleInterface.Request memory _request
+    ) external onlyFromOptimisticOracle {
+        bytes32 depositHash = relayRequestAncillaryData[_ancillaryData];
+        RelayData storage relay = relays[depositHash];
+        require(relay.relayState == RelayState.Pending);
+        relay.relayState = RelayState.Disputed;
+    }
+
+    /**
+     * @notice Callback for settlements, marks relay as ready for finalization if the relay was resolved as valid.
+     * @dev Reverts if relay is uninitialized or already settled.
+     * @dev _timestamp and _identifier are unused because _ancillaryData contains a relay nonce and uniquely
+     * identifies a relay request.
+     * @param _identifier price identifier for relay request.
+     * @param _timestamp timestamp for relay request.
+     * @param _ancillaryData ancillary data for relay request.
+     * @param _request settled relay request params.
+     */
+    function priceSettled(
```

Comment:
Does this mean that a separate settle call in the OO is required for a full lifecycle? What impact does that have on total gas costs?

If this is necessary, why not just do the full settlement in this callback?

Also, I thought our previous model threw out disputes and didn't consider them. It seems like they can now be used to resolve requests? Is that correct? If so, what happens if a dispute comes back in the _middle_ of a follow-up proposal?

Reply:
I guess one argument with not doing the full settlement here is that it will increase the cost of the OO settlement.

Reply:
> Also, I thought our previous model threw out disputes and didn't consider them. It seems like they can now be used to resolve requests? Is that correct? If so, what happens if a dispute comes back in the middle of a follow-up proposal?

The only way for a follow-up proposal to occur is if the original relay gets disputed, due to this line in `relayDeposit`:
```
        require(
            relays[depositHash].relayState == RelayState.Uninitialized ||
                relays[depositHash].relayState == RelayState.Disputed,
            "Pending relay exists"
        );
```

So at this point, either the original dispute can get resolved by the OO, or the follow-up relay can pass liveness and get resolved (by the OO). If the original relay and the follow-up relay have the same params, then the dispute should resolve to 1, and the relay can be settled. If they have the same params, then the payouts are correct and this results in an "early" settlement for the follow-up relay. If the original relay was incorrect, then the dispute will resolve to 0 and the follow-up relay can't be settled yet until it passes liveness.

Reply:
> Does this mean that a separate settle call in the OO is required for a full lifecycle? What impact does that have on total gas costs?

> If this is necessary, why not just do the full settlement in this callback?

We could do this, thoughts?

Reply:
> Does this mean that a separate settle call in the OO is required for a full lifecycle? What impact does that have on total gas costs?

> If this is necessary, why not just do the full settlement in this callback?

Actually, i'm not sure there is an easy way to do this. The OO callsback into `priceSettled` with the ancillary data. We can't really back out the `DepositData` from the `ancillaryData`, as we currently map `ancillaryData => depositHash` and getting `DepositData` from `depositHash` is non-trivial. Without `DepositData` we can't settle easily since that data stores the relay amount for example.